docs/design/metrics-exporter.md

```diff
@@ -0,0 +1,93 @@
+# Custom Resource Metrics Exporter
+
+# Summary
+
+Custom Resource Metrics Exporter provides an insight into the OpenShift Container Storage Cluster (OCS) by aggregating data from various Custom Resources
+and exposing metrics about it. The exposed metrics will be in Prometheus compliant format and can be used to build alerts and charts.
+
+# Motivation
+
+The main motivation behind this exporter is to improve user experience for the consumers of OCS. Currently, cluster state and other custom resource related information can only be found by digging deeper into multiple layers via CLI or Console.
+
+To improve that, this exporter will surface out those details in the form of metrics which can then be consumed by Promethues for analysis as required. Alerts will be created on top of that to notify the user.
+
+# Proposal
+
+The exporter will watch for events on desired Custom Resources and Namespaces and create meaningful metrics with labels. The metric type will be one of Counters, Gauge, Histogram or Summary. Help text and description will be provided along with the metrics for easy understanding.
+
+For example, we will provide RGW health check information based on the CephObjectStore CR. Required parts of the Custom Resource spec is presented below.
+
+```YAML
+apiVersion: ceph.rook.io/v1
+kind: CephObjectStore
+metadata:
+  name: default
+  namespace: openshift-storage
+spec:
+  ...
+  gateway:
+    ...
+    externalRgwEndpoints:
+      - ip: 10.70.56.202
+  ...
+status:
+  bucketStatus:
+    details: >-
+      failed to create object user
+      "rook-ceph-internal-s3-user-checker-9e599ac7-229c-43d4-8e23-40974c9e9177".
+      error code 1 for object store "default": failed to create s3 user: exit
+      status 5
+    health: Failure
+    lastChecked: '2020-07-16T19:20:58Z'
+  info:
+    endpoint: 'http://rook-ceph-rgw-default.openshift-storage:8080'
+  phase: Ready
+```
+
+Name field will be added to label so that the users can uniquely identify the metrics and corelate to historical data. Being an ID, it is unbounded and can have high cardinality.
+
+Namespace field can have many values depending on the number of namespaces in the cluster. To limit this value, only `openshift-storage` will be used by default and users can enable watching more namespaces using flags.
+
+Health field will have a pre-determined set of values. All these values are considered for label.
+
+Based on these data, we create a Promethues Gauge metric `ocs_rgw_health_status` whose value will be 0(healthy), 1(warning) or 2(failure).
+Labels will be added to provide context as described above.
+
+The exposed metrics will look like:
+
+```
+ocs_rgw_health_status {name="default", namespace="openshift-storage",  rgw_endpoints="10.70.56.202", health="Failure"} 2
```

Comment:
There should always be a single endpoint at this stage, the one provided by the Service. So essentially `Status.Info["endpoint"]`.